### ISSUE TITLE
The pickle merge broke some of my queries

### ISSUE THREAD

**Issue by pvalsecc** [geoalchemy/geoalchemy2#191]
Since the commit 93ee16d, a few queries are crashing on my side with a stacktrace looking like that:
```
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/attributes.py", line 242, in __get__
api_1        |     return self.impl.get(instance_state(instance), dict_)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/attributes.py", line 599, in get
api_1        |     value = self.callable_(state, passive)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/strategies.py", line 623, in _load_for_state
api_1        |     return self._emit_lazyload(session, state, ident_key, passive)
api_1        |   File "<string>", line 1, in <lambda>
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/strategies.py", line 747, in _emit_lazyload
api_1        |     result = q(session).params(**params).all()
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/ext/baked.py", line 434, in all
api_1        |     return list(self)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/ext/baked.py", line 336, in __iter__
api_1        |     self.session._autoflush()
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/session.py", line 1434, in _autoflush
api_1        |     self.flush()
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/session.py", line 2254, in flush
api_1        |     self._flush(objects)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/session.py", line 2380, in _flush
api_1        |     transaction.rollback(_capture_exception=True)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/util/langhelpers.py", line 66, in __exit__
api_1        |     compat.reraise(exc_type, exc_value, exc_tb)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/util/compat.py", line 187, in reraise
api_1        |     raise value
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/session.py", line 2344, in _flush
api_1        |     flush_context.execute()
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/unitofwork.py", line 391, in execute
api_1        |     rec.execute(self)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/unitofwork.py", line 556, in execute
api_1        |     uow
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/persistence.py", line 181, in save_obj
api_1        |     mapper, table, insert)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/persistence.py", line 863, in _emit_insert_statements
api_1        |     params)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 948, in execute
api_1        |     return meth(self, multiparams, params)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py", line 269, in _execute_on_connection
api_1        |     return connection._execute_clauseelement(self, multiparams, params)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1053, in _execute_clauseelement
api_1        |     if not self.schema_for_object.is_default else None)
api_1        |   File "<string>", line 1, in <lambda>
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py", line 442, in compile
api_1        |     return self._compiler(dialect, bind=bind, **kw)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py", line 448, in _compiler
api_1        |     return dialect.statement_compiler(dialect, self, **kw)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/compiler.py", line 455, in __init__
api_1        |     Compiled.__init__(self, dialect, statement, **kwargs)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/compiler.py", line 221, in __init__
api_1        |     self.string = self.process(self.statement, **compile_kwargs)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/compiler.py", line 247, in process
api_1        |     return obj._compiler_dispatch(self, **kwargs)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/visitors.py", line 81, in _compiler_dispatch
api_1        |     return meth(self, **kw)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/compiler.py", line 2048, in visit_insert
api_1        |     self, insert_stmt, crud.ISINSERT, **kw)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/crud.py", line 57, in _setup_crud_params
api_1        |     return _get_crud_params(compiler, stmt, **kw)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/crud.py", line 137, in _get_crud_params
api_1        |     _col_bind_name, check_columns, values, kw)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/crud.py", line 269, in _scan_cols
api_1        |     implicit_returning, implicit_return_defaults, values, kw)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/crud.py", line 338, in _append_param_parameter
api_1        |     value = compiler.process(value.self_group(), **kw)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/compiler.py", line 247, in process
api_1        |     return obj._compiler_dispatch(self, **kwargs)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/visitors.py", line 81, in _compiler_dispatch
api_1        |     return meth(self, **kw)
api_1        |   File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/compiler.py", line 933, in visit_function
api_1        |     name = FUNCTIONS.get(func.__class__, func.name + "%(expr)s")
api_1        | TypeError: unsupported operand type(s) for +: '_FunctionGenerator' and 'str'
```
I've added this line to the crashing SQLAlchemy function:

```python
LOG.warning("xxxx " + repr(func) + " - " + repr(func.name))
```

And the output is:
```
xxxx <WKBElement at 0x7efd7807fdc8; 01010000200808000000000000bbdb43410000000014ae3041> - <sqlalchemy.sql.functions._FunctionGenerator object at 0x7efd7809f9a8>
```

Before the offending commit, the output was:
```
xxxx <WKBElement at 0x7f903c04d0b8; 01010000200808000000000000bbdb43410000000014ae3041> - 'ST_GeomFromEWKB'
```

I really don't understand what is happening there and why this commit breaks things like that...

If I checkout ad136d4, everything works nicely.

**Comment by elemoine** [comment]
Sorry to hear that! I'll have a look as soon as I get a chance.

**Comment by elemoine** [comment]
Just took a brief look and I don't understand either. Do you have a small test case reproducing the issue?

**Comment by pvalsecc** [comment]
OK, I found what weird operation I'm doing. Basically I'm duplicating an ORM object and doing some modifications to it. But when duplicating the object, I use `copy.copy` to copy the geometry. 

The code looks like that:
```python
import binascii
from copy import copy
from sqlalchemy import create_engine, Column, Integer
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker
from geoalchemy2 import Geometry, WKBElement

SRID = 2056

engine = create_engine('postgresql://www-data:www-data@localhost:5432/test', echo=True)
Base = declarative_base()
Session = sessionmaker(bind=engine)


class POI(Base):
     __tablename__ = 'pois'
     id = Column(Integer, primary_key=True, autoincrement=True)
     geom = Column(Geometry("POINT", srid=SRID))


def main():
    engine.execute("CREATE EXTENSION IF NOT EXISTS postgis")
    Base.metadata.create_all(engine)

    poi = POI(geom=WKBElement(binascii.unhexlify('01010000200808000000000000bbdb43410000000014ae3041'), srid=SRID, extended=True))
    session = Session()
    session.add(poi)
    session.flush()
    id_ = poi.id

    session.expunge_all()
    back_poi = session.query(POI).get(id_)
    new_poi = POI(geom=copy(back_poi.geom))
    session.add(new_poi)
    session.flush()  # Boom
    session.commit()


main()
```

At the moment of the flush, it goes boom. I guess, in may case, accessing the relation attribute was triggering a flush().

Now, I guess that the `copy` module uses the pickle stuff to copy? If you want, you can not fix that (I can put a work around in my generic code to duplicate an object).

**Comment by pvalsecc** [comment]
Easier to reproduce:
In [1]: from geoalchemy2 import Geometry, WKBElement
In [3]: import binascii
In [4]: e=WKBElement(binascii.unhexlify('01010000200808000000000000bbdb43410000000014ae3041'), srid=2056, extended=True)
In [5]: from copy import copy
In [6]: e2=copy(e)
In [9]: e.name
Out[9]: 'ST_GeomFromEWKB'
In [10]: e2.name
Out[10]: <sqlalchemy.sql.functions._FunctionGenerator at 0x7fe760128198>

**Comment by elemoine** [comment]
We lose the name after unpickling. This is a bug in GeoAlchemy. Thanks for the investigation.

**Comment by elemoine** [comment]
I've created a branch for this issue: https://github.com/geoalchemy/geoalchemy2/compare/master...elemoine:unpickling. I am not fully happy with it, because of the call to `functions.Function.__init__` from the `__setstate__` function. That looks a bit hackish.

Are you able to test this and report whether that works for you or not? Thanks.

**Comment by pvalsecc** [comment]
OK, I'm going to try

**Comment by elemoine** [comment]
Thanks for testing. I'll create a PR based on my branch.

**Comment by elemoine** [comment]
@pvalsecc I merged #192. Tell me if you need a new GeoAlchemy2 release.

**Comment by pvalsecc** [comment]
Cool. Thanks Eric.
No rush on my side. I've put in place a work around.
CU
